src/scratchOrgInfoApi.ts

```diff
@@ -0,0 +1,495 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+// @salesforce
+import { Optional } from '@salesforce/ts-types';
+import { env, Duration, upperFirst } from '@salesforce/kit';
+import { ensureString, getString } from '@salesforce/ts-types';
+
+// Thirdparty
+import { RecordResult, OAuth2Options } from 'jsforce';
+import { retry, retryDecorator, RetryError } from 'ts-retry-promise';
+
+// Local
+import { Org } from './org';
+import { Logger } from './logger';
+import mapKeys from './util/mapKeys';
+import { AuthInfo } from './authInfo';
+import { Messages } from './messages';
+import { SfdxError } from './sfdxError';
+import { SfdcUrl } from './util/sfdcUrl';
+import { Lifecycle } from './lifecycleEvents';
+import { MyDomainResolver } from './status/myDomainResolver';
+import SettingsGenerator from './scratchOrgSettingsGenerator';
+
+export interface ScratchOrgInfo {
+  AdminEmail?: string;
+  readonly CreatedDate?: string;
+  ConnectedAppCallbackUrl?: string;
+  ConnectedAppConsumerKey?: string;
+  Country?: string;
+  Description?: string;
+  DurationDays?: string;
+  Edition?: string;
+  readonly ErrorCode?: string;
+  readonly ExpirationDate?: string;
+  Features?: string;
+  HasSampleData?: boolean;
+  readonly Id?: string;
+  Language?: string;
+  LoginUrl: string;
+  readonly Name?: string;
+  Namespace?: string;
+  OrgName?: string;
+  Release?: 'Current' | 'Previous' | 'Preview';
+  readonly ScratchOrg?: string;
+  SourceOrg?: string;
+  readonly AuthCode: string;
+  Snapshot: string;
+  readonly Status: 'New' | 'Creating' | 'Active' | 'Error' | 'Deleted';
+  readonly SignupEmail: string;
+  readonly SignupUsername: string;
+  readonly SignupInstance: string;
+  Username: string;
+  settings: Optional<Record<string, unknown>>;
+  objectSettings: Optional<Record<string, unknown>>;
+  orgPreferences: Optional<{
+    enabled: string[];
+    disabled: string[];
+  }>;
+}
+
+export interface JsForceError extends Error {
+  errorCode: string;
+  fields: string[];
+}
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/core', 'scratchOrgInfoApi');
+
+const checkScratchOrgInfoForErrors = (orgInfo: ScratchOrgInfo, hubUsername: Optional<string>, logger: Logger) => {
+  if (orgInfo.Status === 'Active') {
+    return orgInfo;
+  }
+  if (orgInfo.Status === 'Error' && orgInfo.ErrorCode) {
+    throw new SfdxError(messages.getMessage('signupFailed', [orgInfo.ErrorCode]), 'RemoteOrgSignupFailed');
```

Comment:
this should have all the error codes.  See the current version in toolbelt.

They're missing from the messages.json on this PR

Reply:
I mean these: https://github.com/salesforcecli/toolbelt/blob/main/messages/scratchorg_error_codes.json

this was fixed here after you started this work because Judy noticed the error codes were missing.  https://github.com/salesforcecli/toolbelt/blob/main/src/lib/org/scratchOrgErrorCodes.ts